Add interop functionality

schwartz-concordium opened the pull request:
## Purpose

Added interop functionality to deserialize schema, entrypoint messages and events.

Covers first two subtask of
https://github.com/Concordium/concordium-net-sdk/issues/68

https://concordium.atlassian.net/browse/CBW-1508

## Changes

- Added functionality to deserialization from module schema
  - Contract entrypoint messages
  - Contract events
  - Module schema

## Checklist

- [x] My code follows the style of this project.
- [ ] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [x] (If necessary) I have updated the CHANGELOG.

schwartz-concordium commented:
Validate build order between C# and Rust before merge.

limemloh commented on src/Types/ContractEvent.cs:
```diff
@@ -9,4 +12,31 @@ public sealed record ContractEvent(byte[] Bytes)
+    /// <param name="contractName">Contract name.</param>
+    /// <returns><see cref="Bytes"/> deserialized.</returns>
+    /// <exception cref="InteropBindingException">Thrown when event wasn't able to be deserialized from schema.</exception>
+    public string? GetDeserializeEvent(
```
**Suggestion:** Make signature like `TryParse` and indicate in the name that this can fail (`Try` prefix).

Since this is parsing the bytes in the event, it should probably follow the same style as `TryParse`, if you would like we can also provide a function which throws.

schwartz-concordium replied:
Then penalty of throwing the exception is already present since it is thrown from the interops operation. I have had this discussion with some from UFT team when introducing alike in the CCD scan, and we concluded best C# practice was to have an exception thrown. 
I believe in must cases the user would like to have the exception if it fails, instead of a silent swallow with a `Try`.

limemloh commented on src/Types/VersionedModuleSchema.cs:
```diff
@@ -0,0 +1,38 @@
+    /// <summary>
+    /// Module schema in hexadecimal.
+    /// </summary>
+    public string Schema { get; }
```
**Suggestion:** Store the schema as raw bytes instead of hex encoding.

limemloh commented on src/Types/ContractEvent.cs:
```diff
@@ -9,4 +12,31 @@ public sealed record ContractEvent(byte[] Bytes)
+    /// <param name="contractName">Contract name.</param>
+    /// <returns><see cref="Bytes"/> deserialized.</returns>
+    /// <exception cref="InteropBindingException">Thrown when event wasn't able to be deserialized from schema.</exception>
+    public string? GetDeserializeEvent(
```
**Suggestion:** The output should indicate that this is JSON instead of just `string`, maybe [`Utf8JsonReader`](https://learn.microsoft.com/en-us/dotnet/standard/serialization/system-text-json/use-utf8jsonreader) is a good match, since it won't actually deserialize the JSON string and the user can bring their own class for this.

schwartz-concordium replied:
Sum up of discussion p2p. It would be nice to show output is json but since `Utf8JsonReader` is a `ref struct` it is limited how it can be used. We discussed `JsonNode` or `JsonDocument` but in the end it will be a overhead.

We will write it clearly in the documentation and then let it be up to the user to parse the result as efficient as they like.

limemloh commented on src/Types/ContractTraceElement.cs:
```diff
@@ -71,7 +73,55 @@ public sealed record Updated(
+    /// <summary>
+    /// Deserialize message from <see cref="schema"/>.
+    /// </summary>
+    /// <param name="schema">Module schema in hexadecimal.</param>
```
```suggestion
    /// <param name="schema">Module schema.</param>
```

limemloh commented on src/Interop/InteropBinding.cs:
```diff
@@ -0,0 +1,157 @@
+    /// <param name="schema">Module schema in hexadecimal</param>
+    /// <param name="schemaVersion">Optional schema version if present from module</param>
+    /// <returns>Module schema in a human interpretable form</returns>
+    internal static string? SchemaDisplay(string schema, ModuleSchemaVersion? schemaVersion)
```
When is the return value `null`? Please add this to the documentation.

schwartz-concordium replied:
Changed return type to not be null

schwartz-concordium commented on src/Concordium.Sdk.csproj:
```diff
@@ -68,6 +73,38 @@
+    <Exec Command="cargo build --manifest-path=../rust-bindings/Cargo.toml --release" />
+  </Target>
+
+  <ItemGroup>
+    <Content Include="..\rust-bindings\Cargo.toml">
+      <Link>rust-bindings\Cargo.toml</Link>
+    </Content>
+    <Content Include="..\rust-bindings\src\lib.rs">
+      <Link>rust-bindings\src\lib.rs</Link>
+    </Content>
+    <Content Include="..\rust-bindings\test-data\cis2_wCCD_sub">
+      <Link>rust-bindings\test-data\cis2_wCCD_sub</Link>
+    </Content>
+  </ItemGroup>
```
Mistake

schwartz-concordium commented on src/Types/ContractEvent.cs:
```diff
@@ -9,4 +12,31 @@ public sealed record ContractEvent(byte[] Bytes)
+    /// <exception cref="InteropBindingException">Thrown when event wasn't able to be deserialized from schema.</exception>
+    public string? GetDeserializeEvent(
+        VersionedModuleSchema schema,
+        string contractName
```
Should be able to use just from a Contract Update event, where you don’t want to map a string from “contract_name.entrypoint” -> “contract_name” -> “init_contract_name” -> “contract_name”. The last two steps are just unnecessary.

abizjak commented on .github/workflows/build-test-format.yml:
```diff
@@ -7,22 +7,52 @@ on:
+        components: rustfmt, clippy
+
+    - name: Format Rust Project
+      run: cargo fmt --manifest-path=./rust-bindings/Cargo.toml --check
```
Yes, and please use the same nightly formatter version for that.

schwartz-concordium replied:
I have added the `rustfmt.toml`. 


In `concordium-base` pipelines format and clippy runs in two different jobs with two different rust version. May I ask why we do this?

abizjak replied:
The stable formatter does not support the unstable options we use. THe main reason here is that rustfmt very very slowly stabilizes anything. The options we are using have been there for years now, but are not stabilized yet.

schwartz-concordium replied:
I have updated the pipeline and added format and clippy to multiple steps. Please verify and resolve if ok.

abizjak commented on rust-bindings/src/lib.rs:
```diff
@@ -0,0 +1,299 @@
+/// If the call succeeded or not.
+///
+/// # Safety
+///
+/// This function is marked as unsafe because it performs operations that are not checked by the
+/// Rust compiler.
```
Usually the safety section should explain how to use it safely.

schwartz-concordium replied:
Sections updated

schwartz-concordium commented on src/Interop/InteropBinding.cs:
```diff
@@ -0,0 +1,157 @@
+    {
+        if (ptr != IntPtr.Zero)
+        {
+            Marshal.FreeHGlobal(ptr);
```
have made a callback approach

schwartz-concordium commented on src/Concordium.Sdk.csproj:
```diff
@@ -68,6 +73,38 @@
+
+  <ItemGroup>
+    <!-- Include librust.dll on Windows -->
+    <None Include="../rust-bindings/target/release/librust_bindings.dll" Condition="'$(IsWindows)'=='true'">
```
Sum up discussion p2p. We will need to have all binaries built for multiple platforms. We will look into making a `nuspec` file which helps us package correctly.

I have made an independent PR which will fix this.
https://github.com/Concordium/concordium-net-sdk/pull/79

abizjak commented on .github/workflows/build-test-format.yml:
```diff
@@ -7,22 +7,70 @@ on:
+    runs-on: ubuntu-latest
+
+    steps:
+    - uses: actions/checkout@v3
```
```suggestion
    - uses: actions/checkout@v4
```

abizjak commented on .github/workflows/build-test-format.yml:
```diff
@@ -7,22 +7,70 @@ on:
+      with:
+        submodules: recursive
+
+    - name: Set up Rust
+      uses: actions-rs/toolchain@v1
+      with:
+        profile: minimal
+        toolchain: ${{ env.RUST_FMT }}
+        override: true
+        components: rustfmt
+
+    - name: Format Rust Project
+      run: cargo fmt --manifest-path=./rust-bindings/Cargo.toml --all --check
```
```suggestion
    - name: Format Rust Project
      run: |
        rustup default ${{ env.RUST_FMT }}
        rustup component add rustfmt
        cargo fmt --manifest-path=./rust-bindings/Cargo.toml --all --check
```

I recommend avoiding actions-rs in new pipelines since it is unmaintained. I recently migrated all of concordium-base jobs away from using these actions. The base ubuntu image has everything we need already.

abizjak commented on .github/workflows/build-test-format.yml:
```diff
@@ -7,22 +7,70 @@ on:
         submodules: recursive
+
+    - name: Set up Rust
+      uses: actions-rs/toolchain@v1
+      with:
+        profile: minimal
+        toolchain: ${{ env.RUST_CLIPPY }}
+        override: true
+        components: clippy
```
```suggestion
      uses: actions-rs/toolchain@v1
      run: |
        rustup default ${{ env.RUST_CLIPPY }}
        rustup component add clippy
```

abizjak commented on rust-bindings/src/lib.rs:
```diff
@@ -0,0 +1,310 @@
+use std::{ffi::CStr, os::raw::c_char};
+
```
```suggestion
```

And let the formatter sort them.

abizjak commented on src/Concordium.Sdk.csproj:
```diff
@@ -45,29 +49,50 @@
       <PrivateAssets>all</PrivateAssets>
     </PackageReference>
     <PackageReference Include="System.Linq.Async" Version="6.0.1" />
+    <PackageReference Include="WebAssembly" Version="1.2.1" />
```
Can you explain what this is for?

schwartz-concordium replied:
We are using it to parse the wasm module and “entering” the custom sections to extract the schema.

abizjak commented on src/Concordium.Sdk.csproj:
```diff
@@ -45,29 +49,50 @@
+      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+    </None>
+    <!-- Include librust.dylib on macOS -->
+    <None Include="../rust-bindings/target/release/librust_bindings.dylib" Condition="'$(IsOSX)'=='true'">
```
Is this x86 or aarch64? Will it depend on what machine this thing is built?

schwartz-concordium replied:
Yes dependent on which platform it is build on the `librust_bindings.*` will be different.

Please note that this is required for local execution. The packaging, which includes all binaries for Linux, iOS, and Windows, will be handled in another PR.

Regarding your query about 'Is this x86 or aarch64,' could you provide further clarification or context? I'm unsure about the specific information you're seeking.

abizjak replied:
There are currently two mac platforms widely used, the older x86-64 intel macs and the newer M{1,2,3} based ones. You can use the x86 code on m1 macs via rosetta but not the other way around.

So my concern is that we package the SDK in the right way so that it's actually usable on both types of platform.

schwartz-concordium replied:
In this PR, which handles packaging, I have made a pipeline which builds on mac os. 
https://github.com/Concordium/concordium-net-sdk/pull/79

Looking at the mac os build step I can see the version used for rust is `1.68-x86_64-apple-darwin`. So from what I understand we are packaging correctly.
https://github.com/Concordium/concordium-net-sdk/actions/runs/6971280381/job/18970984682

I've followed the flow used in the Java SDK release pipeline. If we haven’t encountered any issues related to the Java release I wouldn’t expect any on the .NET.
https://github.com/Concordium/concordium-java-sdk/blob/main/.github/workflows/release.yml#L114

abizjak replied:
The Java flow hasn't really been tested, it's very new.

But it does look that macos-latest is still intel based so it is fine.

Thanks for checking up.

abizjak reviewed (commented):
I'm not a fan of including all of these binary files (modules, schemas) with no documentation on how they were generated, and how they should be updated. Experience suggests this is not maintainable.

So please include that.

If I missed the documentation please point me to it.

limemloh commented on src/Types/ContractTraceElement.cs:
```diff
@@ -71,7 +72,52 @@ public sealed record Updated(
+        var deserialized = new List<Utf8Json>(this.Events.Count);
+        foreach (var contractEvent in this.Events)
+        {
+            var deserializeEvent = contractEvent.GetDeserializeEvent(schema, this.ReceiveName.GetContractName());
```
**Minor thing:** Since `GetContractName` is allocating a new string, it should probably be moved out of the loop.

limemloh commented on src/Types/RejectReason.cs:
```diff
@@ -182,11 +182,22 @@ public sealed record OutOfEnergy : IRejectReason;
+    /// </summary>
+    /// <param name="schema">Versioned module schema.</param>
+    /// <returns><see cref="Parameter"/> deserialized as json uft8 encoded.</returns>
+    /// <exception cref="InteropBindingException">Thrown when message wasn't able to be deserialized form schema.</exception>
```
```suggestion
    /// <exception cref="InteropBindingException">Thrown when message wasn't able to be deserialized using the schema.</exception>
```

limemloh commented on tests/UnitTests/Data/module.schema_embedded.wasm.hex:
**Note (non-blocking):** You seem to switch a bit between using hex encoded and binary files for the test data.
Maybe just have all of them in binary?

schwartz-concordium replied:
Yes that is a good point. I will continue adding only binaries and let those generated in hexadecimal be there. I will write in documentation that one should prefer generating binary test data.

limemloh reviewed (approved):
This looks good to me.

I would prefer not to merge this, until we have the packaging PR done (or nearly done), since we otherwise won't be able to release after this PR is merged.

abizjak commented on README.md:
```diff
@@ -41,6 +41,18 @@ in your project root. It can also be used as a GIT submodule by embedding the cl
+
+The .NET SDK uses Foreign Function Interface (FFI) functions to code written in Rust. The Rust code is located in `./rust-bindings`.
+
+Whenever the .NET SDK is build the Rust code is autimatically compiled. This is done as a `PreBuildEvent` defined in the `Concordium.Sdk.csproj` file. Hence any changes
```
```suggestion
Whenever the .NET SDK is build the Rust code is automatically compiled. This is done as a `PreBuildEvent` defined in the `Concordium.Sdk.csproj` file. Hence any changes
```

abizjak commented on rust-bindings/src/lib.rs:
```diff
@@ -0,0 +1,309 @@
+    Ok(deserialized)
+}
+
+unsafe fn slice_from_ptr<'a, T>(data: *const T, size: usize) -> &'a [T] {
+    std::slice::from_raw_parts(data, size)
+}
```
I agree.

abizjak commented on src/Types/ModuleSchemaVersion.cs:
```diff
@@ -0,0 +1,28 @@
+public enum ModuleSchemaVersion
+{
+    /// <summary>
+    /// Version not known and possible embedded in schema.
```
```suggestion
    /// Version not known and possibly embedded in schema.
```

abizjak commented:
> > I'm not a fan of including all of these binary files (modules, schemas) with no documentation on how they were generated, and how they should be updated. Experience suggests this is not maintainable.
> > So please include that.
> > If I missed the documentation please point me to it.
> 
> I have added a section in the `README.md` about the rust code.
> 
> Are you also referring to test data?

Thanks for including the documentation on the relationship with the Rust code. But I was also talking about all those schema files and modules.

They should be documented on how they are to be updated and what is important about them.

abizjak reviewed (approved):
Please include a description of how the test files are meant to be maintained.

schwartz-concordium commented:
> > > I'm not a fan of including all of these binary files (modules, schemas) with no documentation on how they were generated, and how they should be updated. Experience suggests this is not maintainable.
> > > So please include that.
> > > If I missed the documentation please point me to it.
> > 
> > 
> > I have added a section in the `README.md` about the rust code.
> > Are you also referring to test data?
> 
> Thanks for including the documentation on the relationship with the Rust code. But I was also talking about all those schema files and modules.
> 
> They should be documented on how they are to be updated and what is important about them.

I have added a section now.
